src/annotator/config/index.js

```diff
@@ -1,53 +1,240 @@
+ * @prop {ConfigDefinition} [assetRoot]
+ * @prop {ConfigDefinition} [branding]
+ * @prop {ConfigDefinition} [clientUrl]
+ * @prop {ConfigDefinition} [enableExperimentalNewNoteButton]
+ * @prop {ConfigDefinition} [experimental]
+ * @prop {ConfigDefinition} [group]
+ * @prop {ConfigDefinition} [focus]
+ * @prop {ConfigDefinition} [theme]
+ * @prop {ConfigDefinition} [usernameUrl]
+ * @prop {ConfigDefinition} [onLayoutChange]
+ * @prop {ConfigDefinition} [openSidebar]
+ * @prop {ConfigDefinition} [query]
+ * @prop {ConfigDefinition} [requestConfigFromFrame]
+ * @prop {ConfigDefinition} [services]
+ * @prop {ConfigDefinition} [showHighlights]
+ * @prop {ConfigDefinition} [notebookAppUrl]
+ * @prop {ConfigDefinition} [sidebarAppUrl]
+ * @prop {ConfigDefinition} [subFrameIdentifier]
+ * @prop {ConfigDefinition} [externalContainerSelector]
+ */
+
 /**
  * Reads the Hypothesis configuration from the environment.
- *
- * @param {Window} window_ - The Window object to read config from.
  */
-export default function configFrom(window_) {
-  const settings = settingsFrom(window_);
-  return {
-    annotations: settings.annotations,
-    // URL where client assets are served from. Used when injecting the client
-    // into child iframes.
-    assetRoot: settings.hostPageSetting('assetRoot', {
-      allowInBrowserExt: true,
-    }),
-    branding: settings.hostPageSetting('branding'),
-    // URL of the client's boot script. Used when injecting the client into
-    // child iframes.
-    clientUrl: settings.clientUrl,
-    enableExperimentalNewNoteButton: settings.hostPageSetting(
-      'enableExperimentalNewNoteButton'
-    ),
-    experimental: settings.hostPageSetting('experimental', {
-      defaultValue: {},
-    }),
-    group: settings.group,
-    focus: settings.hostPageSetting('focus'),
-    theme: settings.hostPageSetting('theme'),
-    usernameUrl: settings.hostPageSetting('usernameUrl'),
-    onLayoutChange: settings.hostPageSetting('onLayoutChange'),
-    openSidebar: settings.hostPageSetting('openSidebar', {
-      allowInBrowserExt: true,
-      // Coerce value to a boolean because it may come from via as a string
-      coerce: toBoolean,
-    }),
-    query: settings.query,
-    requestConfigFromFrame: settings.hostPageSetting('requestConfigFromFrame'),
-    services: settings.hostPageSetting('services'),
-    showHighlights: settings.showHighlights,
-    notebookAppUrl: settings.notebookAppUrl,
-    sidebarAppUrl: settings.sidebarAppUrl,
-    // Subframe identifier given when a frame is being embedded into
-    // by a top level client
-    subFrameIdentifier: settings.hostPageSetting('subFrameIdentifier', {
-      allowInBrowserExt: true,
-    }),
-    externalContainerSelector: settings.hostPageSetting(
-      'externalContainerSelector'
-    ),
-  };
+export default class Config {
+  /**
+   * @param {Window} window_ - The Window object to read config from.
+   */
+  constructor(window_) {
+    this.settings = settingsFrom(window_);
```

Comment:
Not related to this set of changes, but man, the return type/value of `settingsFrom` is not very intuitive, is it? Can we easily answer the question: "What does `settingsFrom` return"?

Reply:
no, I was hoping to refactor settingsFrom just be the thing that found the verbatim settings from original sources.  I'm not sure what it should look like. My plain was to largly ignore it and build out index to be more readab

> Good progress here and I'm glad to see that we're slowing untangling this Gordian knot. I spent a while looking at this this morning, various comments inline. My two higher-level areas of consideration are:
> 
> 1. Class design, lifecycle, API
> 2. Definition of which configuration properties belong with which contexts
> 
> ## 1: Class design
> I am finding myself wanting to poke at whether or not this should be designed as a class. There are two factors that make me squint.
> 
> ### 1a: Lifecycle: When are configuration property values calculated?
> There is a lack of clarity over when configuration property values are computed. In the previous implementation (`settingsFrom` function), the settings would be parsed/read/defaulted/whatever at the time the function was called. There was some underlying ambiguity about whether configuration property values (i.e. settings) could change after application startup/boot , but there was an underlying sense of “get me the settings as they are right now.”
> 
> Designing this as a class implies a lifecycle and that the `getConfig` method may be invoked immediately, later, repeatedly or never. That leaves some confusion over what I’d get, exactly, if I invoked `myConfigObj.getConfig(‘notebook’)` at any specific point in time. Would I get the settings as they are exactly right now, or would I get settings as they were at the time of object instantiation? Or a mix?
> 
> Right now, the reality appears to be a potentially-confusing mix (FWIW, a lot of this confusion was pre-existing, but the refactor to a class intensifies it). `isBrowserExtension` is a setting that’s read by the constructor, so that would never change. However, there are a number of configuration properties (settings) that are based on a JSON object or the current URL, e.g.—these could conceivably change after object instantiation. But `isBrowserExtension` is frozen after instantiation…
> 
> ### 1b: Intended public API
> The API and use cases for this logic. These two lines in `annotator/index` again suggest that a class design might be awkward:
> 
> ```
> const config = new Config(window);
> const annotatorConfig = config.getConfig(‘annotator’);
> ```
> 
> OTOH, later, the same `config` object is used to get sidebar configuration (`config.getConfig('sidebar'))`
> 
> What advantages does a class design give us here? Are there any other useful public services or attributes that `config` would ever provide other than this “get me a config” method? How much “work” or recomputing is saved by using a class? What is the intended public API surface?
> 
> ## 2: Defining per-context configuration properties
> I also had some ideas about how we might define what configuration properties apply to which contexts. One option could be defining the configuration property keys in one place (within or outside of the class, or wherever…). This would obviate the need to have to have separate _methods_ for composing these:
> 
> ```
> const configurationContexts = {}; // Or this.configurationContexts
> 
> // Default context
> configurationContexts.all = [
>   'annotations',
>   'assetRoot',
>   'branding',
>   // ...
> ];
> 
> // Annotator takes all configuration properties
> configurationContexts.annotator = configurationContexts.all;
> 
> // Sidebar takes all configuration properties
> configurationContexts.sidebar = configurationContexts.all;
> 
> // Notebook takes all configuration properties except `annotations`
> configurationContexts.notebook = configurationContexts.all.filter(configProperty => configProperty !== 'annotations');
> ```
> 
> `all` could be the default configuration returned by the `getConfig` method:
> 
> `getConfig(appContext = 'all')`
> 
> Finally, I think we could simplify the `@typedef` for the `ConfigDefinitionMap` so that we don't need to maintain it over time (comments inline).


Thanks for take a close look at this. I think with the class, its less relevant to have one now. I initially did in because the config was processed at ctor time as to not process it more than once and the filterwas was just done when you called get(), now its all done at the time you call getConfig so we really don't need to keep an instance. I don't really think of classes as lifecycle but that may be more Java/C++ minded where classes are just used for everything. In JS I can see it being more of an implication of lifecycle as we can just use modules to encapsulate like-things. I'll rework this into functions. 

I avoided making extra variables of confg keys (contexts) as to not have too many places where we re-defind lists of config names. It just gets lengthy to maintain, but I'll give it a try and it it seems more clear, then perhaps its worth it.